mmcai-teardown.sh

```diff
@@ -235,20 +248,38 @@ if $remove_cluster_resources; then
         departments.mmc.ai
     '
 
+    log "Requesting CRD deletion..."
     kubectl delete crd $cluster_resource_crds --ignore-not-found &
     cluster_resource_crds_removed=$!
 
     if $force_if_remove_cluster_resources; then
+        log "Force removing cluster resources..."
         for cluster_resource_crd in $cluster_resource_crds; do
-            until [ -z "$(kubectl get crd $cluster_resource_crd --ignore-not-found)" ]; do
+            log "Removing $cluster_resource_crds resources..."
+
+            if ! get_crd_output=$(kubectl get crd $cluster_resource_crd --ignore-not-found); then
+                error_or_found=true
+            elif [ -z "$get_crd_output" ]; then
+                error_or_found=false
+            else
+                error_or_found=true
+            fi
+
+            while $error_or_found; do
```

Comment:
Can this loop terminate if `$(kubectl get crd $cluster_resource_crd --ignore-not-found)` keeps returning an error? I can imagine this working OK if, for example, we delete all the crds in a certain namespace, and then [ -z "$get_crd_output" ] triggers and the variable turns false -- but is it possible for the kubectl to keep giving us some kind of ERROR, thereby propagating the loop?

Reply:
It's possible that this is an infinite loop. I considered a loop limit but then it gets more complicated. Maybe this can be simplified:
```bash
    if $force_if_remove_cluster_resources; then
        log "Force removing cluster resources..."
        for cluster_resource_crd in $cluster_resource_crds; do
            log "Removing $cluster_resource_crds resources..."

            tolerable_errors=10
            if ! get_crd_output=$(kubectl get crd $cluster_resource_crd --ignore-not-found); then
                log_bad "Error getting CRD $cluster_resource_crd other than CRD not found."
                error_or_found=true
                tolerable_errors=$((tolerable_errors - 1))
                log "Tolerable errors remaining for $cluster_resource_crd: $tolerable_errors. Waiting 5 seconds..."
                sleep 5
            elif ; then
                error_or_found=false
            else
                error_or_found=true
            fi

            while [ $tolerable_errors -ge 0 ] && $error_or_found; do
                namespaces=$(kubectl get namespaces -o custom-columns=:.metadata.name)
                for namespace in $namespaces; do
                    if ! get_crd_output=$(kubectl get crd $cluster_resource_crd --ignore-not-found); then
                        log_bad "Error getting CRD $cluster_resource_crd other than CRD not found."
                        error_or_found=true
                        tolerable_errors=$((tolerable_errors - 1))
                        if [ $tolerable_errors -ge 0 ]; then
                            log "Tolerable errors remaining for $cluster_resource_crd: $tolerable_errors. Waiting 5 seconds..."
                            sleep 5
                        else
                            log_bad "Giving up removing CRD $cluster_resource_crd."
                            break
                        fi
                    elif [ -z "$get_crd_output" ]; then
                        error_or_found=false
                        break
                    else
                        error_or_found=true
                        # This should work for cluster-wide resources.
                        if ! resources=$(kubectl get -n $namespace $cluster_resource_crd -o custom-columns=:.metadata.name); then
                            log_bad "Error getting $cluster_resource_crd resources in namespace $namespace."
                            tolerable_errors=$((tolerable_errors - 1))
                            if [ $tolerable_errors -ge 0 ]; then
                                log "Tolerable errors remaining for $cluster_resource_crd: $tolerable_errors. Waiting 5 seconds..."
                                sleep 5
                            else
                                log_bad "Giving up removing CRD $cluster_resource_crd."
                                break
                            fi
                        elif ! [ -z "$resources" ]; then
                            kubectl patch $cluster_resource_crd -n $namespace $resources --type json --patch='[{ "op": "remove", "path": "/metadata/finalizers" }]' || true
                        fi
                    fi
                done
            done
        done
    fi
```

Reply:
Oh dear. Let's, uh, let's leave it as an infinite loop. Maybe let's just add a log_bad if an uncaught error occurs